### DISCUSSION TITLE
Janus Support For multiple tenants with different schema graph with in a key space

### DISCUSSION THREAD

**Discussion by Noushath** [JanusGraph/janusgraph#3563] [Q&A]
Hi Team, 

As part of using Janus graph for a multi tenant scenario would like to know how we can support multiple schemas from different tenants in a given key space. And we would like to make sure as we do reads on graph a tenant is only able to see graph for his data as per his schema.

tenant 1 : person vertex schema(name:string,salary:number,yearsofexperience:number)
tenant2: person vertex schema(name:string,salary:number,age:number)
Could see some options in DSE as mentioned in [here](https://www.experoinc.com/expero-resources/multi-tenant-applications-in-datastax-graph).

Thanks,
Noushath

**Comment by FlorianHockmann** [comment]
You can simply create completely different graphs in JanusGraph and each graph gets its own schema. So for example `tenant1_graph` for tenant 1 and `tenant2_graph` for tenant 2.

If you want to create new graphs dynamically instead of creating custom config for each by hand, then you should look into [the `ConfiguredGraphFactory`](https://docs.janusgraph.org/operations/configured-graph-factory/) and [_Dynamic Graphs_](https://docs.janusgraph.org/operations/dynamic-graphs/). This allows you to create a template config that will be used to create graphs dynamically based on that config, e.g., each graph gets its own CQL keyspace in Cassandra / Scylla, but apart from that all settings are the same.

We're using that also for a multi scenario at my company. We have the same schema for every tenant, but since the schema needs to be applied to each graph individually, using different schemas for different graphs shouldn't be a problem.

**Reply by Noushath** [reply]
@FlorianHockmann ,If we have separate graphs it would mean, we need to have different traversal on each graph for each client as we read/write data for each tenant. Wouldn't this be an overhead to our backend shared microservices?

**Reply by FlorianHockmann** [reply]
Yes, you need different graph traversal sources for each tenant. How else do you want to solve this? If you have one graph traversal source that can access data from all tenants, then you don't have any separation.
But creating a graph traversal source for each graph isn't difficult with dynamic graphs as they are created automatically on the server with the alias `graph1_traversal` [as described here](https://docs.janusgraph.org/operations/dynamic-graphs/#dynamic-graph-and-traversal-bindings). If your using the name (or ID) of a tenant as the graph name, then you can easily create graph traversal sources in your micro services for each tenant.

You could of course also use the `PartitionStrategy` as described in the blog post you linked, but that still requires you to create a separate graph traversal source configured with that strategy for a specific tenant.
It however makes it difficult to use different schemas per tenant as the schema is defined on a per-graph level. You could basically only create a big schema for all tenants, but only use parts of it for some tenants.

**Reply by Noushath** [reply]
Thanks @FlorianHockmann , we will try this and let you know if we are seeing any problems.
